Plugins/Sycl/src/Utilities/QueueWrapper.cpp

```diff
@@ -0,0 +1,96 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Plugins/Sycl/Utilities/QueueWrapper.hpp"
+
+#include "Acts/Plugins/Sycl/Utilities/DeviceSelector.hpp"
+
+#include <string>
+
+#include <CL/sycl.hpp>
+
+namespace Acts::Sycl {
+
+QueueWrapper::QueueWrapper(const std::string& deviceNameSubtring) {
+  initialize(deviceNameSubtring);
+}
+
+QueueWrapper::QueueWrapper(QueueWrapper&& parent) noexcept
+    : m_queue(parent.m_queue), m_ownsQueue(parent.m_ownsQueue) {
+  parent.m_queue = nullptr;
+  parent.m_ownsQueue = false;
+}
+
+QueueWrapper::QueueWrapper(const QueueWrapper& other) {
+  m_queue = other.m_queue;
+  m_ownsQueue = false;
+}
+
+QueueWrapper::~QueueWrapper() {
+  if (m_ownsQueue && (m_queue != nullptr)) {
+    delete m_queue;
+  }
+};
+
+QueueWrapper& QueueWrapper::operator=(QueueWrapper&& rhs) noexcept {
+  // Check whether we have to do anything
+  if (this == &rhs) {
+    return *this;
+  }
+
+  // Destroy this queue,
+  if (m_ownsQueue && (m_queue != nullptr)) {
+    delete m_queue;
+  }
+
+  // Perform the move
+  m_queue = rhs.m_queue;
+  m_ownsQueue = rhs.m_ownsQueue;
+  rhs.m_queue = nullptr;
+  rhs.m_ownsQueue = false;
+
+  // Return this object.
+  return *this;
+};
+
+QueueWrapper& QueueWrapper::operator=(const QueueWrapper& other) {
+  m_queue = other.m_queue;
+  m_ownsQueue = false;
```

Comment:
This might actually be good to fix still...

You must put a `this == &other` check into this operator as well. Otherwise self-assignments would behave in a **very** broken way. (Leaking memory...)

Reply:
Can I resolve this?

Reply:
Very true.